versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/codecs/BsonToEntityConverter.java

```diff
@@ -0,0 +1,158 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.bson.BsonReader;
+import org.bson.BsonType;
+
+import com.dremio.nessie.versioned.store.Entity;
+import com.dremio.nessie.versioned.store.mongodb.MongoDbConstants;
+
+/**
+ * Converter for serializing a BSON object to an Entity.
+ */
+public class BsonToEntityConverter {
+  private static final String MONGO_ID_NAME = "_id";
+
+  /**
+   * Deserializes a Entity attributes from the BSON stream.
+   *
+   * @param reader the reader to deserialize the Entity attributes from.
+   * @return the Entity attributes.
+   */
+  public static Map<String, Entity> read(BsonReader reader) {
+    final BsonType type = reader.getCurrentBsonType();
+    if (BsonType.DOCUMENT != type) {
+      throw new UnsupportedOperationException(
+          String.format("BSON serialized data must be a document at the root, type is %s",type));
+    }
+
+    return readDocument(reader);
+  }
+
+  /**
+   * Read a BSON document and deserialize to associated Entity types.
+   * @param reader the reader to deserialize the Entities from.
+   * @return The deserialized collection of entities with their names.
+   */
+  private static Map<String, Entity> readDocument(BsonReader reader) {
+    reader.readStartDocument();
+    final Map<String, Entity> attributes = new HashMap<>();
+
+    while (BsonType.END_OF_DOCUMENT != reader.readBsonType()) {
+      final String name = reader.readName();
+      if (name.equals(MONGO_ID_NAME)) {
+        reader.skipValue();
+        continue;
+      }
+
+      switch (reader.getCurrentBsonType()) {
+        case DOCUMENT:
+          attributes.put(name, Entity.ofMap(readDocument(reader)));
+          break;
+        case ARRAY:
+          attributes.put(name, readArray(reader));
+          break;
+        case BOOLEAN:
+          attributes.put(name, Entity.ofBoolean(reader.readBoolean()));
+          break;
+        case STRING: {
+          final String value = reader.readString();
+          if (value.startsWith(MongoDbConstants.STRING_PREFIX)) {
+            attributes.put(name, Entity.ofString(value.substring(1)));
+          } else {
+            attributes.put(name, Entity.ofNumber(value.substring(1)));
+          }
+          break;
+        }
+        case BINARY:
+          attributes.put(name, Entity.ofBinary(reader.readBinaryData().getData()));
+          break;
+        default:
+          throw new UnsupportedOperationException(
+            String.format("Unsupported BSON type: %s", reader.getCurrentBsonType().name()));
+      }
+    }
+
+    reader.readEndDocument();
+    return attributes;
+  }
+
+  /**
+   * Read a BSON array and deserialize to associated Entity types.
+   * @param reader the reader to deserialize the Entities from.
+   * @return The deserialized list-type Entity, either a LIST or STRING_SET.
+   */
+  private static Entity readArray(BsonReader reader) {
+    reader.readStartArray();
+    if (reader.readBoolean()) {
```

Comment:
I think we should remove this struct design and get rid of string sets from entity types. I don't believe we currently use them. Can you post a patch doing that and then clean this patch up?